rfcs/20230420-kernel-cache/README.md

```diff
@@ -0,0 +1,292 @@
+# Reducing Primitive Creation Overhead for Dynamic Shapes
+
+## Motivation
+
+oneDNN API requires users to specify tensors dimensions to create a primitive,
+which makes the primitive dependent on the tensors's shapes. Primitive creation is
+an expensive operation, this is why oneDNN has a primitive cache to reduce the
+overhead. However, the primitive cache can only help if the tensors shapes do not
+change between workload iterations otherwise the primitives have to be created for
+the new shapes which incurs additional overhead. The goal of this RFC is to propose
+a solution to reduce that overhead.
+
+## Background
+
+There is a reason why the oneDNN API requires users to specify tensors dimensions
+to create a primitive. The library may use the information to generate a kernel
+specifically for the given shapes to get the best performance.
+
+However, the information may not always be used by the library meaning that the
+generated kernels can be used for different shapes, which makes requirement to
+always specify tensors dimensions is overly strict.
+
+There was an attempt in the past to relax the requirement by introducing so-called
+runtime dimensions support. The idea was to allow users specify a placeholder
+(`DNNL_RUNTIME_DIM_VAL`) instead of the actual dimensions. The general rule for
+the feature was that the more actual dimensions users provide the better
+performance they can get.
+
+It's obvious that in order to support runtime dimensions the kernels must not
+depend on the shapes. Or there should be different kernels that are specific to
```

Comment:
I guess I implied a different definition of the dependency. What I meant is that if we can use the same kernel for different shapes (1 or more dimensions) then it means the kernel doesn't depend on the shape otherwise it does depend on the shape and we cannot re-use it. This is what the "must not" means here.